### ISSUE TITLE
Learning the two-moons distribution with a normalizing flow

### ISSUE THREAD

**Issue by EiffL** [astrodeepnet/sbi_experiments#2]
@Justinezgh here are some examples I have lying around of building a nomalizing flow in jax, and  training it on the two moons distribution:

- [1] full implementation of a NF in JAX+flax but it is kind of outdated: https://github.com/EiffL/jax-nf (see this notebook in particular https://github.com/EiffL/jax-nf/blob/master/notebooks/Vanilla-NVP.ipynb)

- [2] notebook with a NF implementaion in JAX+haiku: https://github.com/EiffL/Quarks2CosmosDataChallenge/blob/main/notebooks/PartII-GenerativeModels-Solution.ipynb (see `Step III: Latent Normalizing Flow` and ignore all the rest, it just shows you how to build a NF with jax and haiku)

So I would say, you can try to rewrite a small notebook, using 1 as an example for how to generate examples from the two moons dataset, and 2 for an example of a slightly better implementation using haiku

Learning objectives:
 - Learn how to implement a Normalizing Flow
 - Get familiar with TensorFlow Probability distributions
 - Get familiar with Jax and DeepMind's Haiku

**Comment by Justinezgh** [comment]
I can't use @jax.jit for the get_batch function (from this notebook : https://colab.research.google.com/drive/1t4DaL02o31OCOFifDaQS2B1f_QN5-iRq?usp=sharing ), when I use it I get this error : 'IndexError: tuple index out of range'

**Comment by EiffL** [comment]
Could you try the following?
```python
@jax.jit
def get_batch(batch_size, seed):
  batch = get_two_moons(sigma= 0.05).sample(batch_size, seed=seed)
  score = jax.vmap(jax.grad(two_moons.log_prob))(batch)
  return batch, score
```

Maybe it's an issue coming from the fact that you build the distribution outside of the jitted function

**Comment by EiffL** [comment]
So, @Justinezgh, I think you are already pretty much all setup to start some fun research and experiments, so I want to show you some preliminary work on this stuff we did with @b-remy last year. 

We were testing a technique called denoising score matching to learn the score field (not the distribution itself), and we did some tests against what a conventional Normalizing Flow could achieve.  Here is a relevant plot:
![image](https://user-images.githubusercontent.com/861591/141869043-f1d64b97-6f08-407f-8ca9-b719e28f7b41.png)
(from this notebook: https://github.com/b-remy/score-estimation-comparison/blob/normalizing_flows/notebooks/NF-DAE-SN-comparison.ipynb)

It shows that when training a Normalizing Flow just for density estimation, the score field can go all wonky. Also, if you think about the change of variable formula in a Normalizing Flow, the score will have two terms, one that comes from the inverse  mapping, and one that comes from the Jacobian determinant. For a RealNVP, @b-remy also made this plot: 
![image](https://user-images.githubusercontent.com/861591/141869488-066cf958-279f-4745-a814-fedaf3f512f9.png)
(https://github.com/b-remy/score-estimation-comparison/blob/normalizing_flows/notebooks/NFlows_where_come_from_the_failures.ipynb)
which shows that the determinant part seems be responsible for most of the bad behavior, it probably implies that the particular shape of the RealNVP determinant is not very regular.

**Comment by EiffL** [comment]
This makes me think we can take as a first angle of attack is to check that for a given choice of normalizing flow architecture, the log density is indeed correctly continuously  differentiable. And thinking about the log determinant term is probably a good idea.

You can also have a look at one of the seminal papers on score matching: https://www.cs.helsinki.fi/u/ahyvarin/papers/JMLR05.pdf

**Comment by Justinezgh** [comment]
Impact of the nb of coupling layers on the score field 
 : https://colab.research.google.com/drive/1H0Q_hgb0Yjtqvyg9RKeqTvt5lSZBNiap?usp=sharing

**Comment by Justinezgh** [comment]
Same but with Neural Spline Flows : https://colab.research.google.com/drive/1IFDmsNUTsHIjQpjnXKIAG3PUeyx6NLux?usp=sharing

**Comment by EiffL** [comment]
```python
batch_size=512

@jax.jit
def get_batch(seed):
  two_moons = get_two_moons(sigma= 0.05)
  batch = two_moons.sample(batch_size, seed=seed)
  return batch
```

simple fix for the jitting of of get_batch, removing the batch_size argument

**Comment by EiffL** [comment]
@Justinezgh this is all super interesting. Two comments:

- [ ] Have you checked (at least theoretically) that the log prob of a normalizing flow using a realNVP is at least twice differentiable? The leaky-relu for instance shouldn't be, and I think it actually has 0 second order gradients almost everywhere. Which could explain why we are having difficulties training on the score.

- [ ] Can you try to learn the score field with a simple regression network instead of a Normalizing Flow, so directly training a function s_\theta(x) to learn the score field, with a dense neural network for instance. If this works well, it means that there is nothing in principle wrong with the score matching loss, and that if there are difficulties, they must come from the particular architecture of the Normalizing Flow.

**Comment by b-remy** [comment]
Hi @Justinezgh , note that if it is more convenient, you can also keep the `batch_size` argument by specifying to `jit` that it is a static argument.

```python
import functools

@functools.partial(jax.jit, static_argnums=(1,))
def get_batch(seed, batch_size):
  two_moons = get_two_moons(sigma= 0.05)
  batch = two_moons.sample(batch_size, seed=seed)
  return batch
```

**Comment by EiffL** [comment]
Sorry, I was too curious.... I quickly tried to train a regression network under a score matching loss to make sure things were not crazy. And it seems to work pretty well:
![image](https://user-images.githubusercontent.com/861591/142941933-294039b0-b866-4538-b893-09888821a351.png)

The loss function nicely goes to zero instead of jumping around as in the NF examples.
![image](https://user-images.githubusercontent.com/861591/142941994-71e4b5fa-e3d2-4e59-bad1-871626b07573.png)

## Training the grads of the NN

For fun I also tried to train the same network, but making it output just a scalar and constraining its gradients, and that doesnt train at all:
![image](https://user-images.githubusercontent.com/861591/142942966-f1cb90d6-ec5a-4d35-8e83-38d0fd1e63f0.png)

![image](https://user-images.githubusercontent.com/861591/142942978-2f14c78c-320a-4573-a10d-7ef41fba4ab9.png)

## Training the grads of the NN with a C\infty neural network

And for even more fun, I tried training the same model again by constraining the grads, replacing relu activation by a `sin` function, as proposed in https://arxiv.org/abs/2006.09661

And BAM! By magic it works \o/
![image](https://user-images.githubusercontent.com/861591/142944163-430623b1-4fe5-4258-b9d7-d1e918687542.png)
(note: for the background on the right I use exp( scalar output of the network ))

And training goes super easily:
![image](https://user-images.githubusercontent.com/861591/142944374-b117878e-621f-45b8-898e-6bc06eff4a85.png)


=> All codes available here: https://github.com/astrodeepnet/sbi_experiments/pull/4

**Comment by Justinezgh** [comment]
https://colab.research.google.com/drive/1OnL56FPKzinJrnL16xFdYXcBiKSPsOmy?usp=sharing :)

**Comment by EiffL** [comment]
Ahaha yep ^^ sorry, this had been bugging me all afternoon and was dying to try, it's pretty fun stuff :-)

**Comment by EiffL** [comment]
So, the next logical step is to build a NF that is by construction Cinfty.

@b-remy reminded me of this paper: https://arxiv.org/pdf/2110.00351.pdf where they actually propose a coupling layer that should be continuously differentiable, to place in a RealNVP. Probably worthwhile to take a look.

**Comment by Justinezgh** [comment]
So just to see, I tried to use the sin activation function for the NN of the coupling layer : 

![image](https://user-images.githubusercontent.com/72011736/143019679-030b458c-3550-45d2-90c1-5a80e59679f9.png)

![image](https://user-images.githubusercontent.com/72011736/143019735-f5d2f0db-2953-45d0-958d-1d292736aae7.png)

Notebook : https://colab.research.google.com/drive/1ZU-w76vJ81-PArB9vr1x9fi7qpZ1AOnu?usp=sharing

**Comment by EiffL** [comment]
interesting interesting yeah, it doesn't seem to help directly :-/ 

So here is what they say in section 4 of (2110.00351):
![image](https://user-images.githubusercontent.com/861591/143025862-dcb66583-7212-4655-aa01-1db39d4403aa.png)
![image](https://user-images.githubusercontent.com/861591/143025909-6041eb89-4335-4914-98cd-479f6f902304.png)

So what they are saying is that with an affine coupling layer, you lose expressivity in the gradients of the log p. And they also say that the Neural Spline Flows have poor gradients because they are only C1.

So I think we could try the following: Using a C\infty coupling layer, and training under the Score Matching loss (because whether or not the flow can train under the SM loss on its own will tell us if the model is well adapted). 

Alternatively, it might be possible to use a MAF instead of a RealNVP, because it's possible that if the masked autoencoder in the MAF layer is Cinfty, so will be the flow

**Comment by EiffL** [comment]
(just for the record, what I said there about MAF was stupid, you still have an Affine Coupling with a MAF)

**Comment by Justinezgh** [comment]
Just to be sure, this is this function that we want to place in a RealNVP ?

![image](https://user-images.githubusercontent.com/72011736/143133035-41b6b808-15b7-4357-ae8d-2df881e27e10.png)

If yes, do we want to use f to define the shift and the scale part ? Because both shift and scale are R^d -> R^(D-d) so do we have to do some kind of projection for f(x) ? Like we define f(x_i) := (1-c).((g(x_i)-g(0))/(...)) + c.x_i so f(x) \in R^d and then we project in R^(D-d) ?

**Comment by EiffL** [comment]
That's a good question. And yes that's the coupling we might want to use, **instead** of an affine coupling.

So you don't generate shift and scale parameters, instead you generate these a,b,c parameters which are the outputs of some neural network which takes R^d inputs and return R^(D-d) outputs, and the function g is a bijection in R^(D-d).

You can have a look at how the Spline flows work, it's a bit different, but illustrates how a parametrisation can deviate from affine.

**Comment by EiffL** [comment]
It may not be 100% trivial because I think you would have to define a TFP bijector to implement the mapping f. It shouldn't be too difficult, but will take a bit of coding.

Ah, and there is another approach we could take i think.... We could use a ffjord, and there is one easily usable in the TF version of TFP (so not in Jax unfortunately). I **think** if the ode function is sufficiently smooth, so is the ode flow.

**Comment by b-remy** [comment]
+1 I was also thinking that Continuous Normalizing Flows (the flow of transformations being continuous here) such as Neural ODE ([1806.07366](https://arxiv.org/pdf/1806.07366.pdf)) or FFJORD ([1810.01367](https://arxiv.org/pdf/1810.01367.pdf)) would be an interesting approach to look at in parallel!

**Comment by Justinezgh** [comment]
I'm not sure that the function f(x) = (1-c)((g(x)-g(0))/(g(1)-g(0)))+cx has an analytical inverse

**Comment by EiffL** [comment]
hummmmmmmmmmmmm  that sounds surprising

**Comment by EiffL** [comment]
ok, maybe the exp is  hard to find an analytical inv ^^' the monomial should be easier, and otherwise we could  impllement a generic purpose inverse function, with gradients computed by the  implici function theorem.

**Comment by EiffL** [comment]
;-) wink wink @b-remy

**Comment by b-remy** [comment]
I've been looking at [ffjord](https://www.tensorflow.org/probability/examples/FFJORD_Demo), and we can indeed observe that working with a Continuous Normalizing Flow, which makes smooth transformations,  yields a smoother score function!

![image](https://user-images.githubusercontent.com/30293694/144769234-bec9b198-7e6f-487e-b3f6-a9b600307928.png)
https://colab.research.google.com/drive/1nCs0UH8CfToW6L4ZNehzERBdIx84Eg6k?usp=sharing

Here I used maximum likelihood only, no score matching loss because I have not figured out how to implement it with tensorflow yet...

Maybe we should open a specific issue dedicated to ODE flows, to discuss different loss functions or how the gradients are actually computed. And maybe consider implementing a JAX version because taking gradients, or computing vjp, is not as easy in TF :-)

**Comment by EiffL** [comment]
Yep @b-remy agreed, we can open a separate issue to discuss using an ODE flow for this :-) 
We can use this as a plan B, if plan A of using custom coupling layer doesn't work. 

@Justinezgh do you have some news on building an invertible coupling? If not analytically possible, we can use an implicit function trick to define the gradients of a numerical inverse. @b-remy already has experience with this, it's a little bit more involved, but if we don;t analytic inverses it should work.

**Comment by Justinezgh** [comment]
I think the best I can do is rho(x) = x**2 :/
https://colab.research.google.com/drive/1kRA4ReFryVqFJfLxwtL7nXg-Uwsn1sal?usp=sharing

**Comment by EiffL** [comment]
\o/ x**3 should work for our purposes! and maybe x**2 is actually enough... we just need one more order of smoothness than the typical affine coupling.

Let's see what this gives us  in practice in a bijector :-)

**Comment by EiffL** [comment]
BTW, do you know this: https://github.com/SymJAX/SymJAX ;-) 

Really awesome that you used sympy for solving this BTW!

**Comment by EiffL** [comment]
Any luck with implementing a bijector? Don't hesitate if you have questions ;-)

**Comment by Justinezgh** [comment]
I have "some" bugs :D 

https://colab.research.google.com/drive/1cmtlXbH-xX7s7m7MtiL4DWyD_UriSoIg?usp=sharing

When I try to train the NF I have this error (1024 is the batch size)  : 'ValueError: The arguments to _cofactor_solve must have shapes a=[..., m, m] and b=[..., m, m]; got a=(1024, 1, 1024, 1) and b=(1024, 1, 1024, 1)' 

So I tried with batch_size = 1 and I noticed that the loss is NaN. I tried the same thing with an easier bijector Exp() and I have the same pb for the loss. So I tried to print() everything in the NN to get a,b,c  and for some reason the initialization part fails

**Comment by EiffL** [comment]
so several things, 

- batch_size=1 is for sure going to give you unstable traininig
- An exp bijector is almost surely going to explode during trainining, so you might have nans very quickly
- This way of computing the log det is probably going to be instable: `jnp.log(jnp.abs(jnp.linalg.det(jax.jacfwd(f, argnums = 0)(x,self.a, self.b, self.c))))` is there an analytic jacobian for this bijector ?
- The reason why it fails for your batch size >1 is probably because of the jax.jacfwd, you need to make sure you compute a batched jacobian matrix, it should give you [1024, 1, 1], not [1024,1,1024,1]

**Comment by EiffL** [comment]
did it help ^^' ?

**Comment by Justinezgh** [comment]
yup ! 
I computed the gradients with Sympy  : https://colab.research.google.com/drive/1URrqY8TVf0EbtO2DHpjqEnR4jIvs2j-P?usp=sharing
I don't know if it's faster to have the Jacobian for both f and f^-1 or to use the fact that forward_log_det_jacobian is the negative of inverse_log_det_jacobian, evaluated at `g^{-1}(y).

And so now I'm dealing with a new pb :D 
![image](https://user-images.githubusercontent.com/72011736/145408043-8c9a75b3-5d43-4626-9ab0-6c96ec1c5d85.png)
Just have to find a way to have x \in [-1/2a+b, 1/2a+b]

**Comment by EiffL** [comment]
This looks good Justine, but I didn't quite get your point about x in a given range... To keep things simple for now, can we define a flow that remains between (0,1) ?
